sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1950,16 +1950,39 @@ class Analyzer(override val catalogManager: CatalogManager)
       // Replace the index with the corresponding expression in aggregateExpressions. The index is
       // a 1-base position of aggregateExpressions, which is output columns (select expression)
       case Aggregate(groups, aggs, child) if aggs.forall(_.resolved) &&
-        groups.exists(_.isInstanceOf[UnresolvedOrdinal]) =>
-        val newGroups = groups.map {
-          case u @ UnresolvedOrdinal(index) if index > 0 && index <= aggs.size =>
-            aggs(index - 1)
-          case ordinal @ UnresolvedOrdinal(index) =>
-            throw QueryCompilationErrors.groupByPositionRangeError(index, aggs.size, ordinal)
-          case o => o
-        }
+        groups.exists(containUnresolvedOrdinal) =>
+        val newGroups = groups.map((resolveGroupByExpressionOrdinal(_, aggs)))
         Aggregate(newGroups, aggs, child)
     }
+
+    private def containUnresolvedOrdinal(e: Expression): Boolean = e match {
+      case _: UnresolvedOrdinal => true
+      case Cube(_, groupByExprs) => groupByExprs.exists(containUnresolvedOrdinal)
```

Comment:
how about `g: GroupingSet => g.groupByExprs.exists(containUnresolvedOrdinal)`

Reply:
BTW we should refine the name. `GroupingSets extends GroupingSet` looks weird.

Reply:
> BTW we should refine the name. `GroupingSets extends GroupingSet` looks weird.

Renaming to `GroupingAnalytic`

Reply:
> how about `g: GroupingSet => g.groupByExprs.exists(containUnresolvedOrdinal)`

If we remove the strange grammer `group by 1, 2 grouping sets((a, b), ())`, I think we can do like this since

Reply:
After your code merged https://github.com/apache/spark/pull/32022, then update this pr's code